Implement PioneerTestKit and PioneerEngineExecutionResults (#6)

Bukama opened the pull request:
```
Up to now projects tests depends on copied internal JUnit Jupiter classes. In the past this was necessary but we always wanted to get rid of them. With the upgrade to JUnit 5.4.2 this is now possible, as the JUnit Jupiter team provides an public `EngineKit` as an API to those classes.

So with this PR, Pioneer gets its own `TestKit` and `EngineExectionResults` to wrap those of the JUnit Jupiter test engine as our tests always execute the `pioneer test engine`. This wrapper also provides more flexibility adding new functionality and/or easier access to things like the published `ReportEntries` of the executed test method.

PR: #218
closed #6 

ci skip-release
```
---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

Michael1993 commented on src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:
```diff
@@ -83,26 +83,18 @@ void repeatedAnnotation_logEachKeyValuePairAsIndividualEntry() {
 		@Test
 		@ReportEntry(key = "Crow2", value = "While I pondered weak and weary")
 		void explicitKey() {
+			System.out.println("HI THERE! " + executionCount.getAndIncrement());
```
Why is this here?

Bukama replied:
Nicolai added it for basic tests after creating the basic wrapper after updating so see if it works. And I just didn't think about it anymore. Will remove it

Michael1993 commented on src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:
```diff
@@ -83,26 +83,18 @@ void repeatedAnnotation_logEachKeyValuePairAsIndividualEntry() {
 
 	static class ReportEntriesTest {
 
+		private static AtomicInteger executionCount = new AtomicInteger(0);
```
Don't really see why this is here. Just for testing?

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,156 @@
+import java.util.Map;
+import java.util.Optional;
+
+import org.junit.platform.engine.TestDescriptor;
```
Test descriptor is 'unused' in the strictest sense because it is only mentioned in comments. I think we should use fully-qualified names in docs.

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,156 @@
+import org.junit.platform.testkit.engine.Events;
+
+/**
+ * Pioneer own class to handle {@link EngineExecutionResults} of the JUnit-Jupiter-Engine.
```
'Pioneers' to show possession. Sounds better to me.

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,156 @@
+				.tests()
+				.reportingEntryPublished()
+				.stream()
+				.map(executionEvent -> executionEvent.getPayload(org.junit.platform.engine.reporting.ReportEntry.class))
```
You could import this (in which case the import is used over the `ReportEntry` in the same package).

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,156 @@
+				.map(executionEvent -> executionEvent.getPayload(org.junit.platform.engine.reporting.ReportEntry.class))
+				.filter(Optional::isPresent)
+				.map(Optional::get)
+				.map(org.junit.platform.engine.reporting.ReportEntry::getKeyValuePairs)
```
You could import this.

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java:
```diff
@@ -0,0 +1,40 @@
+	public static PioneerEngineExecutionResults getExecutionResults(Class<?> testClass, String testMethodName) {
+		PioneerEngineExecutionResults results = new PioneerEngineExecutionResults(testClass, testMethodName);
+
+		return results;
```
You could inline this (i.e.: `return new Pioneer...`).

Bukama replied:
I know, but I normally don't do it, to have a better debug possibility and the costs for this variable are quite less in my opinion.

Also in this specific case (and the one below) the PioneerTestKit is a wrapper for future improving. We will then have to split it again.

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java:
```diff
@@ -0,0 +1,40 @@
+	public static PioneerEngineExecutionResults getExecutionResults(Class<?> testClass) {
+		PioneerEngineExecutionResults results = new PioneerEngineExecutionResults(testClass);
+
+		return results;
```
You could inline this (i.e.: `return new Pioneer...`).

Bukama replied:
See above

Bukama commented:
Those rebased and forced pushes on master really screw me up. Regardless if I rebase a branch on updated master or merge master into it. All changed files from master are shown in the PR....

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,155 @@
+ */
+public class PioneerEngineExecutionResults {
+
+	EngineExecutionResults executionResults;
```
Is this package-private on purpose?

Bukama replied:
Yes I always use package private instead of private to acces fields and methods from tests

Michael1993 replied:
But in this case - do you access it? I'm not seeing the tests for this class.

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,155 @@
+	public PioneerEngineExecutionResults(Class<?> testClass) {
+		executionResults = EngineTestKit
+				.engine("junit-jupiter")
+				.selectors(DiscoverySelectors.selectClass(testClass))
```
Consider static importing `DiscoverySelectors` methods. Would that be more readable?

Bukama replied:
Personally I like regular imports in such cases more as it's clear that this messages does not belong to the class. But if the majority wants to use static imports I can switch.

nicolaiparlog replied:
I like static imports in general, but I prefer them in situations where a somewhat universal API is used that is frequently seen. `DiscoverySelectors::selectClass` doesn't quite qualify, but I'm still ok, either way.

Bukama replied:
I would like to keep them. So I resolve this conversation.

Michael1993 commented on src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:
```diff
@@ -0,0 +1,155 @@
+	public PioneerEngineExecutionResults(Class<?> testClass, String testMethodName) {
+		executionResults = EngineTestKit
+				.engine("junit-jupiter")
+				.selectors(DiscoverySelectors.selectMethod(testClass, testMethodName))
```
Consider a static import.

Bukama commented:
Moved all extension tests to the new API, but now two tests of the TempDirExtension fail because it seems that they are not started anymore and I can't figure out why :(

Bukama commented:
> Moved all extension tests to the new API, but now two tests of the TempDirExtension fail because it seems that they are not started anymore and I can't figure out why :(

I mean: *NOW* I moved *ALL* extension tests to the new API and removed the superfluous classes.

But still two failing tests and I don't know f*** why 😢 If they would run, this would be mergeable.

Bukama commented:
> > Moved all extension tests to the new API, but now two tests of the TempDirExtension fail because it seems that they are not started anymore and I can't figure out why :(
> 
> I mean: _NOW_ I moved _ALL_ extension tests to the new API and removed the superfluous classes.
> 
> But still two failing tests and I don't know f*** why 😢 If they would run, this would be mergeable.

NAILED them 🥳

nicolaiparlog commented:
Great work, @Bukama! This was tough, but we made it. :tada: Let's hope we get a lot of green lights, so we can merge this.
